Sources/Layout/MessageSizeCalculator.swift

```diff
@@ -219,6 +228,16 @@ open class MessageSizeCalculator: CellSizeCalculator {
 
         return rect.size
     }
+
+    /// - Returns: margin between the Leading edge of the Collection View and the Leading edge of the incoming message bubble.
+    public func incomingMessageLeftMargin() -> CGFloat {
+        return incomingAvatarSize.width + incomingMessagePadding.left + messagesLayout.sectionInset.left
```

Comment:
These values are already taken into account when calling `layoutAvatarView`. We apply the padding to the origin of the `AvatarView`. 

Also, we don't need to consider the `sectionInset` because this is considered when calling `itemWidth` from the `MessagesCollectionViewFlowLayout` object.

Reply:
I don't think the dynamic margin in here is a better way, maybe user don't need this behavior, but add a delegate method in here is ok IMO, @SD10 what do you think? 🤔

Reply:
@zhongwuzw Can you clarify what you mean? I don't want to make it any more dynamic than `outgoing` and `incoming`. And I want all properties to be stored properties. Or else we're heading back to where we were before

Reply:
<img width="372" alt="image" src="https://user-images.githubusercontent.com/5061845/40524649-3167d480-600e-11e8-813e-f585c4f0f431.png">
As screenshot from @Artheyn , because we use the `outgoing` and `incoming`, so users can not dynamic adjust the insets to make text align to the bubble.
I think add delegate for `labelAlignment` is ok, because it's only effect label's internal drawing.

Reply:
Thanks for clarifying @zhongwuzw 💯 Didn't you mention before that users can add the alignment to `NSAttributedString`? I will look into the effects of adding a delegate

Reply:
But also note that the two `Steven` labels above are laid out properly. The bottom label is wrong because the avatar is misplaced

Reply:
> These values are already taken into account when calling layoutAvatarView. We apply the padding to the origin of the AvatarView.

Yeap, it was a way to dynamically compute the labels insets. There is probably a better way.

> But also note that the two Steven labels above are laid out properly. The bottom label is wrong because the avatar is misplaced

In fact, the insets (LabelAlignment) was statically set to `42` before, and I couldn't find any place where it was updated, so I think it cannot be related to the avatar position.

In my opinion we could have a default behaviour which aligns properly the top and bottom labels to the leading/trailing edge of the chat bubble and allow the integrator to update this through delegate like @zhongwuzw says or even subclassing.

Reply:
> In fact, the insets (LabelAlignment) was statically set to 42 before, and I couldn't find any place where it was updated, so I think it cannot be related to the avatar position.

I calculate the frame of the label by using the frame of the `AvatarView`. So if the `AvatarView` frame is wrong, then so is the label.

Reply:
Ok. By inspecting the frames with and without avatar, we can see that the label always take the entire cell view. So there is perhaps an issue in the label frame calculation.

<img width="1392" alt="screen shot 2018-05-25 at 09 54 00" src="https://user-images.githubusercontent.com/6149725/40533127-06d2a706-6002-11e8-8d8d-c2d7e3c98a10.png">

Reply:
@SD10 ,Yeah I mentioned before, remove `textInsets` and `alignment`, and let's user implements these things by `NSAttributedString`.

Reply:
@Artheyn No that is working correctly. Sizing the `NSAttributedString` to fit for a `UILabel` has a huge performance penalty on scrolling.

Reply:
@zhongwuzw Thanks, I'm going to look into this when I get more time. I may have to redo the origin calculations for cells

Reply:
@SD10 , Give you some my thoughts, remove `insetsLabel` and `*LabelAlignment`, instead uses `NSMutableParagraphStyle`'s `firstIndent`、 `tailIndent`、 `alignment `.

Reply:
Ok. Thanks. If I can help any further please let me know.

Reply:
@Artheyn No problem, this is a tricky one to fix because if we apply the `.zero` frame I think we will end up breaking the layout of the other views. This is a really good catch on your part 👍

Reply:
@SD10 ,We need to fix these `.zero` issues, I think set `frame` to `.zero` is ok, do you have any concerns?

Reply:
@zhongwuzw We have to set the frames to `.zero` in another PR. I agree this needs to be done. I can't merge this one because it has other breaking changes that are not in the direction MessageKit is heading.